### ISSUE TITLE
CopyFrom sub-Value to Value

### ISSUE THREAD

**Issue by mercury529** [Tencent/rapidjson#1423]
Hello,

I recently started using rapidjson and am very impressed. Great work. 

I did encounter one situation that may have exposed a potential issue. I have a value in a document. I want to replace the contents of that value with a sub-value that is a part of the destination value (effectively getting rid of every part of the value except the sub-value I want to keep). 

If doing this from a second document, I would use CopyFrom without issue. But I was concerned CopyFrom might not handle the situation properly where rhs is a sub-value of the destination value. So I looked at the implementation:

```c++
GenericValue& CopyFrom(const GenericValue<Encoding, SourceAllocator>& rhs, Allocator& allocator) {
        RAPIDJSON_ASSERT(static_cast<void*>(this) != static_cast<void const*>(&rhs));
        this->~GenericValue();
        new (this) GenericValue(rhs, allocator);
        return *this;
    }
```

If I am not mistaken, this poses a potential risk. The value destructor fires which deallocates memory. This means rhs will no longer be safe to use (although I imagine this will often work, and did work in my sample code, because of the likely short time frame between deallocation and assignment). 

If I am correct about the behavior, I would suggest two things. First, ideally the assert would catch the situation where rhs is part of the destination value. As an alternative to an assert, you could perhaps make a copy of rhs to use when you encounter this corner case. Second, documenting the behavior in the corner case in CopyFrom would be useful. 

For my specific situation, can you recommend a way to achieve what I want in the most efficient way? Two options I see are to copy the entire rhs to a new value and then CopyFrom that or to iterate through the document and erase every value except rhs. Is one more performant than the other with a large document and a large rhs? Is there another alternative that already exists I am missing?

Thanks.

**Comment by pah** [comment]
You are correct, that this corner case is currently broken.  As we cannot easily detect the "nestedness" across mulitple hierarchy levels, I suggest to "move" the original value to a temporary (i.e. nulling out the current value, but keeping the memory alive) and let the temporary's destructor take care of the release afterwards:

```cpp
GenericValue& CopyFrom(const GenericValue<Encoding, SourceAllocator>& rhs, Allocator& allocator) {
        RAPIDJSON_ASSERT(static_cast<void*>(this) != static_cast<void const*>(&rhs));
        GenericValue old(*this); // move old value into temporary to allow copying from sub-hierarchy
        // this->~GenericValue(); // not strictly needed anymore, as "this" got nulled out before
        new (this) GenericValue(rhs, allocator);
        return *this;
        // destructor of "old" will be run here
    }
```

**Comment by pah** [comment]
That said: I'm not sure if we handle this case correctly in the "move" itself.  As this is what you actually want here.  You shouldn't even need to **copy** anything here at all.  But adding this overhead of two moves to every single move would be quite expensive, but you can do the following on the caller side:

```cpp
  Value subHierarchy = top["some"][0]["where"]; // move sub-hierarchy out of document/value
  top["some"] = subHierarchy; // move it back it at a higher hierarchy
```
